lib/evmone/tracing.cpp

```diff
@@ -64,10 +66,103 @@ class HistogramTracer : public Tracer
 public:
     explicit HistogramTracer(std::ostream& out) noexcept : m_out{out} {}
 };
+
+
+class InstructionTracer : public Tracer
+{
+    struct Context
+    {
+        const uint8_t* const code;  ///< Reference to the code being executed.
+        const int64_t start_gas;
+        const char* const* const opcode_names;
+
+        Context(const uint8_t* c, int64_t g, const char* const* n) noexcept
+          : code{c}, start_gas{g}, opcode_names{n}
+        {}
+    };
+
+    std::stack<Context> m_contexts;
+    std::ostream& m_out;  ///< Output stream.
+
+    void output_stack(const Stack& stack)
+    {
+        const auto size = stack.size();
+        m_out << R"(,"stack":[)";
+        for (int i = 0; i < size; ++i)
+        {
+            if (i != 0)
+                m_out << ',';
+            m_out << R"("0x)" << to_string(stack[i], 16) << '"';
+        }
+        m_out << ']';
+    }
+
+    void on_execution_start(
+        evmc_revision rev, const evmc_message& msg, bytes_view code) noexcept override
+    {
+        m_contexts.emplace(code.data(), msg.gas, evmc_get_instruction_names_table(rev));
```

Comment:
Couldn't we assume that all nested executions are on the same revision, and then save `rev` or `opcode_names` once for entire tracer in case `m_contexts` is empty, instead of putting it into each context?

Reply:
I tried this and this requires some special conditions. E.g. we save it when `depth=0`? Then unit tests starting on other depth stops working. So maybe set it when not set? But when to null it then? At exit when `depth=0`?

Reply:
I suggested to look at `m_contexts.empty()`

Reply:
Yes, this works nicely, thanks.
However, I think it is better to do it separately not to modify the HistrogramTracer now.